src/backend/distributed/executor/local_executor.c

```diff
@@ -0,0 +1,579 @@
+ * The returning tuples (if any) is stored in the CitusScanState.
+ *
+ * The function returns totalRowsProcessed.
+ */
+static uint64
+ExecuteLocalTaskList(CitusScanState *node, List *taskList)
+{
+	EState *executorState = ScanStateGetExecutorState(node);
+	ParamListInfo paramListInfo = copyParamList(executorState->es_param_list_info);
+	Oid *paramOids = GetParamListOids(paramListInfo);
+	int numParams = paramListInfo != NULL ? paramListInfo->numParams : 0;
+
+	ListCell *taskCell = NULL;
+
+	uint64 totalRowsProcessed = 0;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		PlannedStmt *localPlan = NULL;
+		int cursorOptions = 0;
+		const char *shardQueryString = task->queryString;
+		Query *shardQuery = ParseQueryString(shardQueryString, paramOids, numParams);
+
+		/*
+		 * We should not consider using CURSOR_OPT_FORCE_DISTRIBUTED in case of
+		 * intermediate results in the query. That'd trigger ExecuteLocalTaskPlan()
+		 * go through the distributed executor, which we do not want since the
+		 * query is already known to be local.
+		 */
+		cursorOptions = 0;
+
+		/*
+		 * Altough the shardQuery is local to this node, we prefer planner()
+		 * over standard_planner(). The primary reason for that is Citus itself
+		 * is not very tolarent standard_planner() calls that doesn't go through
+		 * distributed_planner() because of the way that restriction hooks are
+		 * implemented. So, let planner to call distributed_planner() which
+		 * eventually calls standard_planner().
+		 */
+		localPlan = planner(shardQuery, cursorOptions, paramListInfo);
+
+		LogLocalCommand(shardQueryString);
+
+		totalRowsProcessed += ExecuteLocalTaskPlan(node, localPlan, task->queryString);
+	}
+
+	return totalRowsProcessed;
+}
+
+
+/*
+ * GetParamListOids gets a ParamListInfo and returns an array of Oids
+ * where each element is the Oid of the corresponding Param in the ParamListInfo.
+ *
+ * If paramList is NULL or doesn't have any parameters, the function returns
+ * NULL.
+ */
+static Oid *
+GetParamListOids(ParamListInfo paramList)
+{
+	Oid *paramListTypes = NULL;
+	int numParamIndex = 0;
+
+	if (paramList == NULL || paramList->numParams <= 0)
+	{
+		return NULL;
+	}
+
+	paramListTypes = (Oid *) palloc(sizeof(Oid) * paramList->numParams);
+
+	for (numParamIndex = 0; numParamIndex < paramList->numParams; numParamIndex++)
+	{
+		ParamExternData *param;
+
+		/* give hook a chance in case parameter is dynamic */
+		if (paramList->paramFetch != NULL)
+		{
+#if (PG_VERSION_NUM >= 110000)
+			ParamExternData paramData;
+
+			param =
+				paramList->paramFetch(paramList, numParamIndex + 1, false, &paramData);
+#else
+			param = &paramList->params[numParamIndex];
+			if (!OidIsValid(param->ptype))
+			{
+				paramList->paramFetch(paramList, numParamIndex + 1);
+			}
+#endif
+		}
+		else
+		{
+			param = &paramList->params[numParamIndex];
+		}
+
+		if (param->isnull)
+		{
+			paramListTypes[numParamIndex] = InvalidOid;
+			continue;
```

Comment:
Question: Do we really want to skip at this point? Or, could we ever get a NULL param?